examples/d3d12/src/main.rs

```diff
@@ -0,0 +1,844 @@
+
+    if command_line.use_warp_device {
+        title.push_str(" (WARP)");
+    }
+
+    let hwnd = unsafe {
+        CreateWindowExA(
+            Default::default(),
+            "RustWindowClass",
+            title,
+            WS_OVERLAPPEDWINDOW,
+            CW_USEDEFAULT,
+            CW_USEDEFAULT,
+            window_rect.right - window_rect.left,
+            window_rect.bottom - window_rect.top,
+            None, // no parent window
+            None, // no menus
+            instance,
+            &mut sample as *mut _ as _,
+        )
+    };
+    debug_assert!(hwnd.0 != 0);
+
+    sample.bind_to_window(&hwnd)?;
+
+    unsafe { ShowWindow(hwnd, SW_SHOW) };
+
+    loop {
+        let mut message = MSG::default();
+
+        if unsafe { PeekMessageA(&mut message, None, 0, 0, PM_REMOVE) }.into() {
+            unsafe {
+                TranslateMessage(&mut message);
+                DispatchMessageA(&mut message);
+            }
+
+            if message.message == WM_QUIT {
+                break;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn sample_wndproc<S: DXSample>(sample: &mut S, message: u32, wparam: WPARAM) -> bool {
+    match message {
+        WM_KEYDOWN => {
+            sample.on_key_down(wparam.0 as u8);
+            true
+        }
+        WM_KEYUP => {
+            sample.on_key_up(wparam.0 as u8);
+            true
+        }
+        WM_PAINT => {
+            sample.update();
+            sample.render();
+            true
+        }
+        _ => false,
+    }
+}
+
+#[allow(non_snake_case)]
+#[cfg(target_pointer_width = "32")]
+unsafe fn SetWindowLong(window: HWND, index: WINDOW_LONG_PTR_INDEX, value: isize) -> isize {
+    SetWindowLongA(window, index, value as _) as _
+}
+
+#[allow(non_snake_case)]
+#[cfg(target_pointer_width = "64")]
+unsafe fn SetWindowLong(window: HWND, index: WINDOW_LONG_PTR_INDEX, value: isize) -> isize {
+    SetWindowLongPtrA(window, index, value)
+}
+
+#[allow(non_snake_case)]
+#[cfg(target_pointer_width = "32")]
+unsafe fn GetWindowLong(window: HWND, index: WINDOW_LONG_PTR_INDEX) -> isize {
+    GetWindowLongA(window, index) as _
+}
+
+#[allow(non_snake_case)]
+#[cfg(target_pointer_width = "64")]
+unsafe fn GetWindowLong(window: HWND, index: WINDOW_LONG_PTR_INDEX) -> isize {
+    GetWindowLongPtrA(window, index)
+}
+
+extern "system" fn wndproc<S: DXSample>(
+    window: HWND,
+    message: u32,
+    wparam: WPARAM,
+    lparam: LPARAM,
+) -> LRESULT {
+    let user_data = unsafe { GetWindowLong(window, GWLP_USERDATA) };
+    let sample: Option<&mut S> = match user_data {
+        0 => None,
+        _ => Some(unsafe { transmute(user_data) }),
+    };
+
+    match message {
+        WM_CREATE => {
+            unsafe {
+                let create_struct: &CREATESTRUCTA = transmute(lparam);
+                SetWindowLong(
+                    window,
+                    GWLP_USERDATA,
+                    transmute(create_struct.lpCreateParams),
```

Comment:
Thoughts on this being explicit `as isize` rather than relying on inference?  The rust docs seem to discourage `as _`.  Of course the previous transmute version was also relying on inference, but it was that way because I didn't realize I could use `as isize` here.

Reply:
Ah yes, it was clippy complaining about this:

```rust
warning: transmute from `*mut std::ffi::c_void` to `isize` which could be expressed as a pointer cast instead
   --> examples\d3d12\src\main.rs:179:54
    |
179 |                 SetWindowLong(window, GWLP_USERDATA, transmute(create_struct.lpCreateParams));
    |                                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `create_struct.lpCreateParams as isize`
    |
    = note: `#[warn(clippy::transmutes_expressible_as_ptr_casts)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#transmutes_expressible_as_ptr_casts
```

Through `transmute` is much better here because it does not allow changing the size and potentially truncating pointers from 64-bit to 32-bit. Consider if `lpCreateParams` was a `u32` and pointer sizes are 64-bit (yes, that's not truncation but the easiest way to demonstrate this), transmute disallows this:

```rust
error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
   --> examples\d3d12\src\main.rs:179:54
    |
179 |                 SetWindowLong(window, GWLP_USERDATA, transmute(create_struct.lpCreateParams as u32));
    |                                                      ^^^^^^^^^
    |
    = note: source type: `u32` (32 bits)
    = note: target type: `isize` (64 bits)
```

But a pointer cast like `create_struct.lpCreateParams as u32 as isize` compiles correctly and horribly fails at runtime.

I guess using an explicit `isize` type should be fine to match the function wrappers though (which do the same `as _` casting. Clock does the same `as _` when calling this.

Reply:
I'll follow the other examples here and go with `as _`.